packages/notifier/src/asyncIteratableKit.js

```diff
@@ -0,0 +1,84 @@
+// @ts-check
+// eslint-disable-next-line spaced-comment
+/// <reference types="ses"/>
+
+import { HandledPromise, E } from '@agoric/eventual-send';
+import { makePromiseKit } from '@agoric/promise-kit';
+
+import './types';
+
+/**
+ * This makes a pair of an `updater` and an initial `asyncIteratable`. The
+ * `updater` API is the same as the Notifier's `updater`. In both cases,  calls
+ * to the `updater` produce a stream of non-final values terminated with
+ * either a final success value or a final failure reason. The purpose of the
+ * notifier is to be lossy over the non-final values. The purpose of the
+ * asyncIterableKit is to consume such streams losslessly.
+ *
+ * The initial `asyncIteratable` represents the stream starting with the first
+ * update to the `updater`. It makes async iterators each of which advance
+ * independently starting at that starting point. An async iterator has a
+ * `snapshot()` method which will create a new async iterable capturing the
+ * iterator's current position as its starting point.
+ *
+ * As is conventional, the async iterator is also an async iterable that
+ * produces an async iterator. In this case, it is a new async iterator that
+ * advances independently starting from the current position.
+ *
+ * The internal representation ensure that eements that are no longer
+ * observable are unreachable and can be gc'ed.
+ */
+export const makeAsyncIteratableKit = () => {
+  const makeAsyncIteratable = startE => {
+    return harden({
+      // eslint-disable-next-line no-use-before-define
+      [Symbol.asyncIterator]: () => makeAsyncIterator(startE),
+    });
+  };
+  // To understand the implementation, start with
+  // https://web.archive.org/web/20160404122250/http://wiki.ecmascript.org/doku.php?id=strawman:concurrency#infinite_queue
+  const makeAsyncIterator = startE => {
+    return harden({
+      snapshot: () => makeAsyncIteratable(startE),
```

Comment:
I don’t think `snapshot` can work as advertised (resuming from the point the snapshot was taken). I would like to see the test case.

Reply:
I believe we've verbally clarified why it does work. Not resolving until I write the desired test case.

Reply:
Wow, testing that was a lot of work. Done. Resolving.